### DISCUSSION TITLE
vertex to atoms

### DISCUSSION THREAD

**Discussion by yuqinan** [schrodinger/pymol-open-source#247] [Q&A]
Hi! How can I know which atom/residue an vertex is connected to?

**Comment by JarrettSJohnson** [comment]
> Hi! How can I know which atom/residue an vertex is connected to?

I might need some more specifics regarding this question. How are you obtaining this vertex data?

**Reply by yuqinan** [reply]
Hi! Thank you so much for you reply. For example, in the CGOGL.CPP file, I am wondering for lines of code like this `glVertex3fv(*v); `. Is it possible for me to know which atom/ residue this v belongs to? Or are there any data structure that keep track of information like this?

**Reply by yuqinan** [reply]
To be more specific, how does the pymol code achieve changing the color for one residue in the cartoon representation. Is every single vertex connected to a residue/atom?

**Reply by JarrettSJohnson** [reply]
As far as I know if you change the color for a portion of a mesh even if it's a single residue, the entire mesh is redrawn. However, if the object is pickable, we assign a picking color that maps back to either an atom index or bond index for molecules.

**Reply by yuqinan** [reply]
Thank you for the reply! As I understand, to change the color of a residue, the code will change the `color` attributes for the atom. However, we need to make the object pickable?

**Reply by yuqinan** [reply]
Could you maybe refer a file where this picking color is done? I am having some problems navigating on how to initiate a pickable object. Thanks a lot!

**Reply by JarrettSJohnson** [reply]
Picking logic might be a little tough to understand:
https://github.com/schrodinger/pymol-open-source/blob/master/layer1/Picking.h
https://github.com/schrodinger/pymol-open-source/blob/master/layer1/CGOGL.cpp#L1884

Can I ask though what you're trying to accomplish?

**Reply by yuqinan** [reply]
Thanks! I am trying to implement the same features of modifying a specific residue on a molecule but visualize it with unreal engine. It was kind of difficult to figure out how the atoms positions and the vertexes are related.

**Reply by yuqinan** [reply]
As a follow up, I am wondering if all the object are initiated as non pickable and then becomes pickable later if part of it is commanded to change?

**Reply by JarrettSJohnson** [reply]
Perhaps I don't follow or we may be discussing two completely different things, but pickable molecules by default are pickable. The picking texture (which stores the picking colors) may not have been generated yet if the user hasn't requested picking (via mouse event).

**Reply by yuqinan** [reply]
Oh it makes more sense now! Is it possible to directly set up the picking texture without using the mouse event? For example, what is the best way to initiate a molecule with the first residue colored yellow? I tried directly editing the AtomInfo color of a ObjectMolecule and then render. But it doesn't seem to work.

**Reply by JarrettSJohnson** [reply]
Apologies for the late response, been away a bit for the past week and will be next week as well; just been looking sporadically and I must've missed this.

I haven't tried changing the data inside AtomInfo directly to see if that makes a change; perhaps it may be overridden by another color being used (maybe object-level) or something.

After re-reading the conversation, I may have just been a bit confused. CGOGL.cpp doesn't really hold any information about the atoms/bonds that make them up (except for the picking information I mentioned before); however, if you're looking for the converse way (where the vertex of a CGO is created *from* the atom/residues, you'd find these in the respective `layer2/Rep*` functions where `CoordSet*`s are being read in and atom-level positions and colors are read to generate the 3D form of the molecule.

If you need more details regarding this, feel free to reach out via email: jarrett.johnson@schrodinger.com
